### PR TITLE
Shadowling Update III

### LINKED ISSUE #9012: Shadowlings abilities can be used while not being a genetic shadowling
#8776 added a path for mutant races to change race and an admittedly costly one already existed in brain transplants. 

If a shadowling manages to change race they'll lose most of their big drawbacks while keeping their abilities. If #8776 is used they'll also keep their name (shadowling #XXX) and NODROP equipment, which are both tells but the latter also gives access to a lot of useful extrinsic effects.

Monkeying and unmonkeying will allow the shadowlings armor to drop. Additionally a shadowling can still use his abilities as a monkey (or anything else one might find itself as).

Genetics/Plastic surgery can mask the name.

A shadowling that does all three things (monkeys/unmonkeys, gets plastic surgery, then uses the new green slime reaction) can effectively remove all downsides and tells to being a shadowling while still using all their spell-like abilities.

---

Suggestions:

Shadowlings already gain an ability that allows them to regenerate their genetic shadowling status and drops, so shifting back from a different mutantrace (but not monkeying) is trivial. I think a shadowling going out of there way to re-disguise like that after hatching is a cool enough concept, they just shouldn't be at full power while masked.

* You should probably add checks for shadow/ling on all their post hatch abilities. 

* Consider adding human_req to most shadowling abilities to prevent a light immune shadowmonkey from ventcrawling and ranged stun converting everyone.

* Shadowling abilities that deal with travel should be given nonabstract_req if they aren't given human_req.

* Regenerate Chitin should probably cure more extreme forms of transformation if needed (Monkeying/Sliming/etc)

* Stretch goal: Move the extrinsic effects carried on a shadowlings items to intrinsic either handled as spell like abilities or as part of the shadow/ling race. If you still don't want them to wear pants, there's vars for that (no_equip and nojumpsuit)

---

span class="userdanger" I gleamed these potential issues from code review, it's possible some of these issues in fact aren't an issue and I just read the code wrong /span

### LINKED ISSUE #9067: Shadowlings can lose hatch verb.
A shadowling died, and got cloned while some others incinerated his corpse. He thus lost the hatch verb.
Hatch is also not a spell/proc holder and it probably should be so this can at least be worked around if it occurs again.
@xhuis

### LINKED ISSUE #10822: Borging shadowling thralls does not de-thrall them
Shadowling Thralls  
Security Cyborg-677 PM 

People complained about borgs with shadowling icons :/

### PR DESCRIPTION
**Pull request by ghost**
I know I said I was on hiatus, but I figured I should get this gamemode a bit more polished. I still won't be working on anything else until August.
- Changes flavor text when made into a shadowling or thrall via player panel
- Allows for thrall deconversion via surgery or borging
- Separates communes: thralls have a 5 second cooldown, shadowlings have none
- Thralls can be examined to reveal their status if next to them while they are not wearing a mask
- Thralls now have minor abilities they can use, including night vision
- Hatch and Ascend are now spells rather than procs
- Generic remove_spell proc has been added to minds to make spell management easier code-wise
- Generic get_lumcount proc has been added to turf to return luminosity count
- You can no longer use hatch-exclusive abilities without having the shadowling mutant race
- Some bugs with various abilities have been resolved
- Some abilities have been tweaked
- Enthrall no longer has a cooldown but is still limited to one enthralling at a time
- Flash Freeze has been renamed to Icy Veins
- An enemy communication intercept passage has been added for shadowlings
- Ascendants now deal 60 melee damage (up from 35)
- Ascendants are no longer affected by bombs or the singularity
- Non-shadowlings can no longer use shadowling spells
- Shadowlings no longer take damage from light while shadow walking
- Shadowling spell sounds are properly generalized
- Glacial Blast has been replaced with Lightning Storm, which does basically the same thing
- Black Recuperation now takes 3 seconds (down from 10)
- Adds spell icons (listed below)
- Fixes #10822
- Fixes #9067
- Fixes #9012

**Spell Icons:**

_Annihilate_
annihilate
_Ascend_
ascend
_Collective Hivemind_
collective_mind
_Hivemind Commune/Lesser Commune/Ascendant Commune_
commune
_Drain Life_
drainlife
_Enthrall/Hypnosis_
enthrall
_Glare/Lesser Glare_
glare
_Hatch_
hatch
_Icy Veins_
icyveins
_Lightning Storm_
lightning
_Regenerate Chitin_
regenerate
_Black Recuperation_
revivethrall
_Shadow Walk/Guise/Phase Shift_
shadowwalk
_Blindness Smoke_
smoke
_Sonic Screech_
screech
_Ascendant Transmit_
transmit
_Veil_
veil

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
Thrall deconversion via surgery is pretty extreme. Rev and Gang have a simple mechanism for detection+removal (implant), so does cult (holy water)---with this, there's literally no way to tell aside from forcing surgery on people and seeing if the tumor shows up or not.

**Comment by Xhuis** [comment]
The reason I went with the surgery approach is because of how much more shadowling relies on thrall. A single headrev can kill all the heads on their own, and a single cultist with an armor rune can murderbone. A shadowling with no thralls and active hunters is severely crippled. It's meant to be slow, and it switches up from the traditional "take over cargo and waste all points on implants" strategy thatrev and gang both use.

**Comment by Fox-McCloud** [comment]
I suppose, but it's also really shitty to force-surgery everyone just to find the potential thralls; at least with cult there's an external way of checking.

**Comment by Cebasiz** [comment]
is there one of those small flashlights doctors use to shine on your eyes in medbay somewhere

maybe shadowling eyes react aversely to strong direct light or something

that way you could quickly test people as long as you could capture 80% of the crew, i don't think it would be too bad, if you think that testing via surgery is too hard.

**Comment by WJohn** [comment]
That'd be terrible, you'd have doctors running around flashing everybody in the eyes to test if they're shadowlings and security would be okay with it because it's harmless if they're human, and can save lives by finding shadowlings before they amass too many.

**Comment by Xhuis** [comment]
You can also just take their mask off and examine them.

File: code/game/gamemodes/wizard/wizard.dm [hunk 1]
```diff
@@ -248,6 +248,20 @@
 		qdel(spell_to_remove)
 		mind.spell_list -= spell_to_remove
 
+/datum/mind/proc/add_spell(var/obj/effect/proc_holder/spell/spell) //To add a specific spell to a mind
```
Comment (AnturK) [inline code/game/gamemodes/wizard/wizard.dm]: There already is AddSpell - handling action buttons etc for both minds and mobs.

**Comment by AnturK** [comment]
Something i'm not sure was already fixed : thralls are showing up in glare choice list.

File: code/game/gamemodes/wizard/wizard.dm (hunk 1 above)
Comment (Xhuis) [inline code/game/gamemodes/wizard/wizard.dm]: vOv

**Comment by Xhuis** [comment]
I don't think I can fix that

File: code/game/gamemodes/wizard/wizard.dm (hunk 1 above)
Comment (AnturK) [inline code/game/gamemodes/wizard/wizard.dm]: Remove spell is good though

**Comment by Durkel** [comment]
Could you perhaps add a limit to the amount of thralls you can have before hatching(or lower it if there is one)? Watching jack asses never hatch is tiresome.

**Comment by Robustin** [comment]
Surgery works hand-in-hand with Thralls being identifiable now (remove mask, examine), though the medbay's penlights (make it channeled to avoid vigilante doctors mass-penning anyone in the halls) would be a pretty neat antag-check tool too. 

Surgery strikes a nice balance IMO. Thralls are huge for Slings, they are needed for ascension, protection, conversion, etc. and losing even one can be crippling. You can't have some fast-and-easy deconversion like gangs/rev when it takes a good deal of time+effort to convert someone in the first place. Surgery is a robust system and can either be done quickly in the OR or security can resort to ghetto surgery to try and turn the tide against the thralls without resorting to mass executions.

**Comment by Xhuis** [comment]
Assuming that the check passes, this is merge-ready

**Comment by LaharlMontogmmery** [comment]
Ascendant no longer affected by bombs? Okay, I can get that.
Ascendant not affected by the singulo? C'me oooooon.

Still, i'm okay with this PR.

**Comment by Fox-McCloud** [comment]
If the singulo eats and ascendant it should make it go super singulo and gives it enough energy to become one (from stage 1) instead =p

**Comment by Xhuis** [comment]
I was actually thinking about giving ascendants the ability to eat the singulo, but... naah

**Comment by Xhuis** [comment]
Changed to apcs_list. This is merge-ready

File: code/game/gamemodes/shadowling/shadowling_abilities.dm [hunk 2]
```diff
@@ -60,8 +101,11 @@
 	H.SetLuminosity(blacklistLuminosity) //I hate lightcode for making me do it this way
 
 /obj/effect/proc_holder/spell/aoe_turf/veil/cast(list/targets)
-	usr << "<span class='shadowling'>You silently disable all nearby lights.</span>"
+	if(!shadowling_check(usr))
+		charge_counter = charge_max
+		return
 	for(var/turf/T in targets)
+		usr << "<span class='shadowling'>You silently disable all nearby lights.</span>"
```
Comment (TheDZD) [inline code/game/gamemodes/shadowling/shadowling_abilities.dm]: Isn't this going to cause message spamming now that it's inside of the loop?

File: code/game/gamemodes/shadowling/shadowling_abilities.dm [hunk 3]
```diff
@@ -404,10 +539,12 @@ datum/reagent/shadowling_blindness_smoke //Blinds non-shadowlings, heals shadowl
...
 
 /obj/effect/proc_holder/spell/aoe_turf/unearthly_screech/cast(list/targets)
-	usr.audible_message("<span class='warning'><b>[usr] lets out a horrible scream!</b></span>")
-	playsound(usr.loc, 'sound/effects/screech.ogg', 100, 1)
-
+	if(!shadowling_check(usr))
+		charge_counter = charge_max
+		return
 	for(var/turf/T in targets)
+		usr.audible_message("<span class='warning'><b>[usr] lets out a horrible scream!</b></span>")
```
Comment (TheDZD) [inline code/game/gamemodes/shadowling/shadowling_abilities.dm]: Looks like the same thing will happen here.

File: code/game/gamemodes/shadowling/shadowling_abilities.dm [hunk 4]
```diff
@@ -333,9 +462,12 @@
...
 				if(is_shadow(M))
-					M.mind.spell_list -= /obj/effect/proc_holder/spell/targeted/collective_mind
-					M.mind.current.verbs -= /mob/living/carbon/human/proc/shadowling_hatch //In case a shadowling hasn't hatched
-					M.mind.current.verbs += /mob/living/carbon/human/proc/shadowling_ascendance
+					var/obj/effect/proc_holder/spell/targeted/collective_mind/CM
+					if(CM in M.mind.spell_list)
+						M.mind.spell_list -= CM
+						qdel(CM)
+					M.mind.remove_spell(/obj/effect/proc_holder/spell/targeted/shadowling_hatch)
+					M.mind.add_spell(/obj/effect/proc_holder/spell/targeted/shadowling_hatch)
```
Comment (TheDZD) [inline code/game/gamemodes/shadowling/shadowling_abilities.dm]: This should probably be shadowling_ascend, not shadowling_hatch.

**Comment by Xhuis** [comment]
August has officially begun so I no longer have an excuse not to do this and my other work. I will fix the issues in this PR soon.

**Comment by PatchouliKnowledge** [comment]
How about using a penlight on their eyes to check their thrall status? The penlight as it is doesn't see that much use.

**Comment by Xhuis** [comment]
That was already debated and decided against in favor of examines.

**Comment by Xhuis** [comment]
Commits squashed, issues fixed, conflicts removed, bibbity bop boop please merge this

File: code/game/gamemodes/wizard/wizard.dm (hunk 1 above)
Comment (Jordie0608) [inline code/game/gamemodes/wizard/wizard.dm]: As said, we don't need another proc to add spells, just use this one.
https://github.com/tgstation/-tg-station/blob/master/code/datums/mind.dm#L1451

**Comment by Xhuis** [comment]
Removed the new add_spell proc and replaced its uses with AddSpell. Also added proper spell icons for all spells!

**Comment by Core0verload** [comment]
Update your surgery. Replace '<location = "head">' with '<possible_locs = list("head")>'. And add this code to the file:
```
/datum/surgery/dethrall/can_start(mob/user, mob/living/carbon/target)
	return is_thrall(target)
```
So the surgery will show up only on thralls.

**Comment by Xhuis** [comment]
Starting on everyone is intentional. I don't actually know what you want me to replace.

**Comment by Core0verload** [comment]
>Replace location = "head" with possible_locs = list("head")

Why start on anyone? If you can spot thrall by examining him closely, spotting thrall by looking at his surgical menu isn't that bad.

**Comment by Xhuis** [comment]
Being able to magically tell if someone is eligible for surgery doesn't appeal to me.

**Comment by Xhuis** [comment]
Surgery is updated, I ran a brief test and it works as intended.

**Comment by Core0verload** [comment]
>Ascendants are no longer affected by bombs or the singularity

WAIT WHAT? No, just no.

**Comment by Xhuis** [comment]
That's exactly what ascendants are and are balanced around. They're a round-ender, akin to a loose singularity or Nar-Sie. If a shadowling gets found out at any time, it's extremely hard to become this powerful. At any rate, they can phase through walls and are not affected by space in the first place. As for bombs, they cannot be damaged by _pulse rifles_, so logically bombs would be the same way.

**Comment by Core0verload** [comment]
I killed The Nar-Sie, Geometer of Blood once. It took two BoHs. Lord Singulo isn't immortal too. 

I repeat, immortal beings with no weaknesses are really bad.

**Comment by Xhuis** [comment]
I would be considerably more inclined to listen to your feedback if you didn't personally bash me for not catering to your opinion.

**Comment by Fox-McCloud** [comment]
Not 100% sure about singulo immunity or explosion...but explosion seems fairly reasonable.

Singulo? Eh, might as well, I guess.

**Comment by Jordie0608** [comment]
The immunity sorta makes sense but can always be rebalanced if it turns out to not work well.

**Pull request merged**